### DISCUSSION TITLE
PyMuPDF on Discord

### DISCUSSION THREAD

**Discussion by jamie-lemon** [pymupdf/PyMuPDF#1766] [General]
We now have a dedicated public Discord channel for discussing PyMuPDF!

Please find it here: [#pymupdf](https://discord.gg/TSpYGBW4eq)

We look forward to supporting your questions! ( and who knows perhaps you can give us a few answers too 😉 )

**Comment by oliviaharman** [comment]
Hello (from the discord server), the PDF is still too large to upload, and I am unsure how to upload the script since ipynb files are not supported. Any suggestions?
Thanks.

**Reply by JorjMcKie** [reply]
Hi I am the same guy, just using a synonym.
You could Zip the ipynb and then attach it. Same problem for every executable file.
Otherwise ...

**Comment by JorjMcKie** [comment]
Please simply attach the PDF here, or share a URL or even send it to my private e-mail: jorj.x.mckie@outlook.de.
More than from your code, I am expecting insights from the PDF itself ...

**Reply by oliviaharman** [reply]
I will send it to your email since I am still having some trouble.
Thank you for being so patient with me.

**Reply by JorjMcKie** [reply]
You are welcome - no problem at all.

**Reply by JorjMcKie** [reply]
received the link etc, ... dowloading now

**Reply by JorjMcKie** [reply]
Half a TB - I understand your comment on the size now 😉- problems with any specific image on some page or just all or many?

**Reply by oliviaharman** [reply]
Glad you got to take a look at it! The problem with the inverted colors is on all pages, except for the cover page.

**Comment by JorjMcKie** [comment]
Here is a working script:
```python
import fitz

xrefs = set()  # prevent multiple processing for same image
doc = fitz.open("test.pdf")
for i in range(doc.page_count):  # iterate over pages
    if i > 0 and i % 50 == 0:  # some entertainment messages
        print(f"Processing page {i}")
    for item in doc.get_page_images(i):  # iterate over images on this page
        # Note we do not load any pages - as we would when using page.get_imges()
        xref = item[0]  # the image xref
        if xref in xrefs:  # skip if done earlier already
            continue
        xrefs.add(xref)
        pix = fitz.Pixmap(doc, xref)
        if pix.colorspace == None:  # skip "mask" images
            continue
        if pix.n not in (1, 3):  # if neither Gray nor RGB, convert
            pix = fitz.Pixmap(fitz.csRGB, pix)
        try:  # things may still go wrong
            pix.save(f"page-{i}-{xref}.png")
        except RuntimeError:  # problems may occurs for special Gray pixmaps
            if pix.n == 1:  # convert one more time
                pix = fitz.Pixmap(fitz.csGRAY, pix)
                pix.save(f"page-{i}-{xref}.png")
```
Comments:
Original images are mostly JPEG in CMYK format. As you wrote: storing them as is shows inverted colors. I am avoiding a long analysis and simply convert them all to either Gray or RGB and then save as PNG (smaller anyway).

**Reply by JorjMcKie** [reply]
@oliviaharman For future questions or contributions, please do not hesitate to open your own Discussion item (or issue where applicable), instead of editing this post.

BTW: very nice pictures and PDF!

**Reply by oliviaharman** [reply]
This is great - even more than I expected. Thank you so much for your help. I will not hesitate to reach out. Thanks again

**Reply by JorjMcKie** [reply]
You are welcome!
I am glad to see you enjoy working with PyMuPDF.
